### PR TITLE
NWM_UDS: Fix some issues, cleanups, better PacketHandling

### PR DESCRIPTION
**Pull request by B3n30**
Some cleanups in NWM_UDS:
- Only handle packets if they match the current state
- Fix some issues in Disconnect
- Some code cleanups (e.g. in SendTo)

There are no changes that require a Network Version Bump

I'd like to get a report of most of the multiplayer games to verify that this PR doesn't break anything, before we merge it.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4299)

### CODE REVIEW

**Comment by B3n30** [comment]
List of Games I'd like to get tested/reports on before merging:

Asphalt Assault 3D
Code of Princess
Dragon Ball Extreme Butoden
Dragon Ball Fusions
Dragon Quest Monsters Joker 3 Pro
Dragon Quest Monsters: Terry
Formula 1 2011
Mario Party Star Rush
Monster Hunter 3 Ultimate
Monster Hunter 3G
Monster Hunter 4
Monster Hunter 4 Ultimate
Monster Hunter 4G
Monster Hunter Generations
Monster Hunter X
Monster Hunter XX
New Super Mario Bros. 2
Planet Crashers
Pokemon Mystery Dungeon - Gates to Infinity
Pokemon OR/AS
Pokemon Su/Mo
Pokemon US/UM
Pokemon Virtual Console
Resident Evil Mercenaries 3D
Ridge Racer 3D
Sonic Generations
Super Smash Bros.
Super Street Fighter IV
The Legend of Zelda: Tri Force Heroes

Please note: neither were all of them working prior to that PR, nor do I expect them all to work now

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 1]
```diff
@@ -492,7 +538,15 @@ void HandleDeauthenticationFrame(const Network::WifiPacket& packet) {
         return;
     }
 
-    u16 node_id = node_map[packet.transmitter_address];
+    s32 node_id = node_map[packet.transmitter_address].node_id;
+    bool connected = node_map[packet.transmitter_address].connected;
```
Comment (wwylele) [inline src/core/hle/service/nwm/nwm_uds.cpp]: Do it in one map lookup:
```
auto [node_id, connected] = node_map[packet.transmitter_address];
```

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 2]
```diff
@@ -103,13 +104,16 @@ static std::mutex beacon_mutex;
 
 // Number of beacons to store before we start dropping the old ones.
 // TODO(Subv): Find a more accurate value for this limit.
-constexpr std::size_t MaxBeaconFrames = 15;
+static constexpr std::size_t MaxBeaconFrames = 15;
```
Comment (wwylele) [inline src/core/hle/service/nwm/nwm_uds.cpp]: constexpr implies static

File: src/core/hle/service/nwm/nwm_uds.cpp [hunk 3]
```diff
@@ -103,13 +104,16 @@ static std::mutex beacon_mutex;
...
 
 // List of the last <MaxBeaconFrames> beacons received from the network.
 static std::list<Network::WifiPacket> received_beacons;
 
 // Network node id used when a SecureData packet is addressed to every connected node.
-constexpr u16 BroadcastNetworkNodeId = 0xFFFF;
+static constexpr u16 BroadcastNetworkNodeId = 0xFFFF;
+
+// The Host has always dest_node_id 1
+static constexpr u16 HostDestNodeId = 1;
```
Comment (wwylele) [inline src/core/hle/service/nwm/nwm_uds.cpp]: ditto

**Comment by tywald** [comment]
Some initial local testing, using Appveyor `6295782` msvc build.

**DQMT**
Fixed issue where the host would wait forever for the opponent after a game in 1vs1.

**DQMJ3 Pro**
No issues in 1vs1.

**MHGen**
In the hub:
It seems to be flawless up to 3 players, joining and leaving multiple times posed no issue however when Player 4 joins and the host closes the hub the leadership doesn't pass on to someone else(which it does when there are 3 players in total). After a few seconds the Player 4 crashes with the assert: "Can't be a router if we're not a host". Player 2 and Player 3 only receives "Disconnected" from the game, it carries on as normal after that.

Demonstration: https://www.youtube.com/watch?v=VZ0qnxMgD5g

Logs: [mhgen_logs.zip](https://github.com/citra-emu/citra/files/2446225/mhgen_logs.zip)

**Comment by jroweboy** [comment]
@MojoJojoDojo https://github.com/citra-emu/citra/issues/3709 if you get a chance, could you test with this PR to see if it fixes the issue in 3709?

**Comment by MojoJojoDojo** [comment]
Nope there is still the same issue as before on Fusions (one-way communication in lobby).

**Pull request merged**